test/e2e/node_drain_timeout.go

```diff
@@ -0,0 +1,168 @@
+			WaitForDeploymentAvailableInterval: input.E2EConfig.GetIntervals(specName, "wait-deployment-available"),
+		})
+
+		By("Scale the machinedeployment down to zero. If we didn't have the NodeDrainTimeout duration, the node drain process would block this operator.")
+		nodeDrainTimeoutMachineDeploymentInterval := convertMachineDeploymentDurationToInterval(machineDeployments[0].Spec.Template.Spec.NodeDrainTimeout)
+		for _, md := range machineDeployments {
+			framework.ScaleAndWaitMachineDeployment(ctx, framework.ScaleAndWaitMachineDeploymentInput{
+				ClusterProxy:              input.BootstrapClusterProxy,
+				Cluster:                   cluster,
+				MachineDeployment:         md,
+				WaitForMachineDeployments: nodeDrainTimeoutMachineDeploymentInterval,
+				Replicas:                  0,
+			})
+		}
+
+		By("Deploy workload on the master node. The workload is actually the pods that we deployed above.")
+		framework.DeployWorkloadOnControlplaneNode(ctx, framework.DeployWorkloadOnControlplaneNodeInput{
+			WorkloadCluster:                    workloadClient,
+			ControlPlane:                       controlplane,
+			WaitForDeploymentAvailableInterval: input.E2EConfig.GetIntervals(specName, "wait-deployment-available"),
+		})
+
+		By("Scale down the controlplane of the workload cluster and make sure that nodes running workload can be deleted even the draining process is blocked.")
+		// When we scale down the KCP, controlplane machines are by default deleted one by one, so it requires more time.
+		nodeDrainTimeoutKCPInterval := convertKCPDurationToInternal(controlplane.Spec.NodeDrainTimeout, controlPlaneReplicas)
+		framework.ScaleAndWaitControlPlane(ctx, framework.ScaleAndWaitControlPlaneInput{
+			ClusterProxy:        input.BootstrapClusterProxy,
+			Cluster:             cluster,
+			ControlPlane:        controlplane,
+			Replicas:            1,
+			WaitForControlPlane: nodeDrainTimeoutKCPInterval,
+		})
+
+		By("PASSED!")
+	})
+
+	AfterEach(func() {
+		// Dumps all the resources in the spec namespace, then cleanups the cluster object and the spec namespace itself.
+		dumpSpecResourcesAndCleanup(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder, namespace, cancelWatches, cluster, input.E2EConfig.GetIntervals, input.SkipCleanup)
+	})
+}
+
+func convertDurationToInterval(duration *metav1.Duration, delayRate int) []interface{} {
+	minIntervalDuration := duration.Duration
+	maxIntervalDuration := (duration.Duration + time.Minute*2) * time.Duration(delayRate)
+	intervals := make([]interface{}, 0, 2)
+	intervals = append(intervals, maxIntervalDuration.String(), minIntervalDuration.String())
+	return intervals
+}
+
+func convertMachineDeploymentDurationToInterval(duration *metav1.Duration) []interface{} {
+	return convertDurationToInterval(duration, 1)
+}
+
+func convertKCPDurationToInternal(duration *metav1.Duration, numKCP int) []interface{} {
+	return convertDurationToInterval(duration, numKCP-1)
+}
```

Comment:
Those funcs can be dropped if we move down to one control plane replica, and also convertDurationToInterval can be simplified by removing the `delayRate` parameter